File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # maxLength=0
        

        # for i in range(len(s)):
        #     checkDuplicate=set()
        #     duplicateFound = false
        #     for j in range(i,len(s)):
        #         if s[j] in  checkDuplicate:
        #             maxLength=max(maxLength,j-i)
        #             break
        #         else:
        #             checkDuplicate.add(s[j])
        # return maxLength
        left=0
        check=set()
        longestSubstring=0
        '''
        we keep adding letters to the set and if we there is char , we increment our left pointer
        '''

        for right in range(len(s)):
            while s[right] in check:
                check.remove(s[left])
                left+=1
            check.add(s[right])
            longestSubstring=max(longestSubstring,right-left+1)
        return longestSubstring

  




            

        
            
                
                
            
                
    
                

                   

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int maxSubStringLength = 0;
        HashSet<Character> lookup = new HashSet<>();

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);

            while (lookup.contains(currentChar)) {
                lookup.remove(s.charAt(left));
                left++;
            }

            lookup.add(currentChar);
            maxSubStringLength = Math.max(maxSubStringLength, right - left + 1);
        }

        return maxSubStringLength;
    }
}
